### ISSUE TITLE
When creating a collection, disease term selection sometimes stores the incorrect term

### ISSUE THREAD

**Issue by beforan** [biobankinguk/biobankinguk#580]
## Environment details

- Operating System: unknown
- Browser: unknown
- Directory environment: biobankinguk prod

## Repro steps:

- Create a new collection
- Choose the disease term "Malignant tumor of pancreas (disorder)"
- Fill out the rest of the collection and save
- The saved collection has the disease term "Local recurrence of malignant tumor of pancreas (disorder)"

## Acceptance criteria

- The term stored should match the term selected at creation

## Status

- [ ] Reproduction confirmed

## Further information

From a preliminary investigation into the environment's data, this isn't an "other terms" issue (I initially thought it might be that the user was selecting an "other term" at creation, and then the collection was showing the "canonical" term). In this case, there are two different terms in play: one enabled for gui use (`DisplayOnDirectory`), the other not. So literally a different SNOMED id is stored than the one that was selected.

I do wonder if we're matching incorrectly on lowercase string values - the values in the table are mixed case and so "Malignant" (the selected term) and "[...] malignant" (the stored term). This needs to be investigated.

**Comment by beforan** [comment]
## Initial findings

This is definitely due to case insensitive matching, resulting in finding multiple matches despite the whole term being used, then selecting the incorrect "first" match from those.

This is an issue in production due to the collation in use on the production SQL Server database (allowing case insensitivity)

It's not an issue that appears when running  v4.0.0 (ASP.NET Core version) of the Directory so far, as we target Postgres and the default postgres collation is case sensitive. This however has caused other issues to appear.

## Initial fix

We have a locally working fix that relies on the postgres case sensitive collation to resolve this issue, and have resolved the other issues caused when case insensitivity *is* desirable.

## Better fix

While the above fixes the reported scenario, it's possible that even with case-sensitive matching, we may erroneously pick an incorrect disease status while we use the partial match code for what should be an exact match. If it's not too hard to isolate the times when this should be done, it would be best to provide up to 3 different Disease Status lookup methods:

1. Case insensitive partial match (as used today by Search and other typeahead scenarios)
2. Case sensitive exact match (as should be used by this scenario, and should select a Single result, not the First of potentially many)
3. Optionally case sensitive partial match (if there is a need for this - all scenarios may actually be covered by the above 2)

An alternative fix would be to use SNOMED IDs more in typeahead result submissions, so we match on ID instead of the text value. That would be an even better alternative to method 2 above, but may be more involved in terms of the code change.

Expect a PR shortly with the initial fix; follow up later with a better fix.

**Comment by AndyRae** [comment]
I cannot recreate this (as it's kind of fixed?), and honestly my brain is not in the code to implement the "better fix". 

@beforan - is this a block to deploying to production in your view?

**Comment by beforan** [comment]
> is this a block to deploying to production in your view?

No, the current fix (#586 in conjunction with using case-sensitive collation e.g. by targeting Postgres) is good enough for production release.

I think it's worth keeping this issue / the "better fix" on the backlog (but not required for 4.0 prod milestone) so it can be picked up pro-actively at a better time, or it's not completely blindsiding if the issue does occur with future data.
